### ISSUE TITLE
A reparameterization trick for discrete actions

### ISSUE THREAD

**Issue by xlnwel** [danijar/dreamer#11]
Hi, @danijar 

For discrete actions, have you considered [gumbel-softmax](https://arxiv.org/abs/1611.01144), a reparameterization trick for categorical distributions? Here's my code if you haven't and would like to give it a try. On the other hand, I would be very happy to know the result if you have done some similar experiments before --- I just found that I didn't have so much resources to train on Atari games.

```
class Categorical(Distribution):
    def __init__(self, logits, tau=1):
        self.logits = logits
        self.tau = tau  # tau in Gumbel-Softmax

    def log_prob(self, x):
        return -self.neg_log_prob(x)

    def neg_log_prob(self, x):
        if x.shape.ndims == len(self.logits.shape) and x.shape[-1] == self.logits.shape[-1]:
            # when x is one-hot encoded
            return tf.nn.softmax_cross_entropy_with_logits(labels=x, logits=self.logits)
        else:
            return tf.nn.sparse_softmax_cross_entropy_with_logits(labels=x, logits=self.logits)

    def sample(self, reparameterize=True, hard=True, one_hot=True):
        """
         A differentiable sampling method for categorical distribution
         reference paper: Categorical Reparameterization with Gumbel-Softmax
         original code: https://github.com/ericjang/gumbel-softmax/blob/master/Categorical%20VAE.ipynb
        """
        if reparameterize:
            g = tfd.Gumbel(0, 1).sample(tf.shape(self.logits))
            # Draw a sample from the Gumbel-Softmax distribution
            y = tf.nn.softmax((self.logits + g) / self.tau)
            # draw one-hot encoded sample from the softmax
            if not one_hot:
                y = tf.cast(tf.argmax(y, -1), tf.int32)
            elif hard:
                y_hard = tf.one_hot(tf.argmax(y, -1), self.logits.shape[-1])
                y = tf.stop_gradient(y_hard - y) + y
            assert y.shape.ndims == len(self.logits.shape)
        else:
            y = tfd.Categorical(self.logits).sample()
            assert y.shape.ndims == len(self.logits.shape) - 1
            if one_hot:
                y = tf.one_hot(y, self.logits.shape[-1])
                assert y.shape.ndims == len(self.logits.shape)
                assert y.shape[-1] == self.logits.shape[-1]

        return y

    def entropy(self):
        probs = tf.nn.softmax(self.logits)
        log_probs = tf.math.log(probs + EPSILON)
        entropy = tf.reduce_sum(-probs * log_probs, axis=-1)

        return entropy

    def kl(self, other):
        probs = tf.nn.softmax(self.logits)
        log_probs = tf.math.log(probs)
        other_log_probs = tf.nn.log_softmax(other.logits)
        kl = tf.reduce_sum(probs * (log_probs - other_log_probs), axis=-1)

        return kl

    def mode(self):
        return tf.argmax(self.logits, -1)
```

**Comment by danijar** [comment]
Thanks. Yes, I have tried it and it also works. However, it requires tuning the temperature. I've found straight through gradients to work equally well and they have no hyper parameter.
